### PR TITLE
Implement Distance Utilities

### PR DESCRIPTION
**Pull request by gordonkoehn**
This PR addresses/closes:
- #74 

It implements the most simple, yet flexible way to generate metric and log_probability output files for a single mcmc run, with snakemake rules.

Example:

`snakemake -c 1 ../data/vis00/analysis/MCMC_42-CS_42-T_d_5_123-1000_0.01_0.02_0.0_f_UXR-iT_d_5_123-MC_1.24e-06_0.097_25_0_1-MPC_0.1_0.65_0.25/T_d_5_123/MP3.json
`
Will generate a `MP3.json` file for a `base_tree = T_d_5_123` and the mcmc samples`MCMC_*` in experiment `vis00`

These snakemake rules are very long, will be sheltered in the actual experiment rules - these rules are just the most general

See outputs:
Screenshot 2023-06-20 at 14 20 22

### CODE REVIEW

File: src/pyggdrasil/analyze/_calculation.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                    tree, and log-probability
+        metric :  Callable[[TreeNode, TreeNode], float]
+                     metric to apply to the trees.
+        base_tree : TreeNode
+                    Tree to compare all applicable metrics to.
+
+    Returns:
+        list[int], list[float]
+        Iteration number and results of the metric.
+    """
```
Comment (gordonkoehn) [inline src/pyggdrasil/analyze/_calculation.py:36]: This is one function that allows calculating all [TreeNode, TreeNode] -> float metrics, that is all distances and also `is_true_tree`

Simples design I could think of.

File: src/pyggdrasil/analyze/_calculation.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    Returns of list of tuples of (iteration, tree, log-probability), or False.
+    Goes through entire chain to find all instances of the tree.
+
+    Args:
+        desired_tree : Tree
+        mcmc_samples : McmcRunData
+    Returns:
+        bool or list(tuple[int, Tree, float])
+    """
```
Comment (gordonkoehn) [inline src/pyggdrasil/analyze/_calculation.py:75]: Moved this function from `tree_inference.analyzer` - no in use currently

File: src/pyggdrasil/analyze/_metrics.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    @staticmethod
+    def get(metric: str) -> Callable[[TreeNode, TreeNode], float]:
+        """Return metric function."""
+        return Metrics._METRICS[metric]
+
+    _METRICS: Dict[str, Callable[[TreeNode, TreeNode], float]] = {  # type: ignore
+        "AD": AncestorDescendantSimilarity().calculate,
+        "MP3": MP3Similarity().calculate,
+        "TrueTree": compare_trees,
+    }
```
Comment (gordonkoehn) [inline src/pyggdrasil/analyze/_metrics.py:21]: Easy to call list of metrics, to be appended if we implement more

File: src/pyggdrasil/analyze/_utils.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    """Converts McmcRunData to PureMcmcData.
+
+    Takes a list of MCMCSamples
+    converts it into a xarray easy to plot.
+
+    Args:
+        mcmc_samples : list[MCMCSample] - list of MCMC samples
+    Returns:
+        PureMcmcData
+    """
```
Comment (gordonkoehn) [inline src/pyggdrasil/analyze/_utils.py:25]: Moved over from `tree_inference.analyzer`

File: src/pyggdrasil/analyze/_utils.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+#         """Initialize Analyzer with a Scorer."""
+#         self.scorer = scorer
+#
+#     def analyze(self, mcmc_samples: PureMcmcData) -> dict[str, list[float]]:
+#         """Analyze trees from MCMC runs given a Scorer."""
+#         scores = {name: [] for name in self.scorer.metrics.keys()}
+#         for tree in mcmc_samples.trees:
+#             for name, score in self.scorer.score(tree).items():
+#                 scores[name].append(score)
+#         return scores
```
Comment (gordonkoehn) [inline src/pyggdrasil/analyze/_utils.py:74]: Possible Future design: in case we want to avoid loading the mcmc data files many times and calculate all metrics at once. As you proposed originally. 

Nice design - just wanted to have something simple working here for now.

File: src/pyggdrasil/interface/__init__.py [hunk 6]
```diff
@@ -84,74 +82,3 @@ def append(
         self.iterations = jnp.append(self.iterations, iteration)
         self.trees.append(tree)
         self.log_probabilities = jnp.append(self.log_probabilities, log_probability)
-
-
-@dataclass
-class AugmentedMcmcData(PureMcmcData):
-    """Distance/Similarity and True Tree boolean enhanced MCMC data
-
-    intended to be human-readable and easy to use for plotting"""
```
Comment (gordonkoehn) [inline src/pyggdrasil/interface/__init__.py:93]: Abandoning this class. 

no need to store all these metrics - can just sav them to files directly.

**Review by gordonkoehn — COMMENTED** [review]
Moved already existing `analysis` from `tree_inference.analyzer` to the new `analyzer` module. The `tree_infernece` module was getting so big. 

I did choose to have a simple design for calculating metics and distances, capable to be easily controlled with snakemake rule for s single mcmc run.

If we want to calculate metrics between all samples of a tree or different mcmc chains I need to think of something new. This was to big a task for now. Wanted to get something simple working first.
Please let me know if you have a good idea for a design that is good for calculating distances of different mcmc runs and withing mcmc run distances.

File: src/pyggdrasil/analyze/_metrics.py (hunk 3 above)
Comment (pawel-czyz) [reply src/pyggdrasil/analyze/_metrics.py:21]: Nice!

**Review by pawel-czyz — APPROVED** [review]
Nice! Could we take a look at a short trajectory tomorrow?

File: src/pyggdrasil/analyze/_utils.py (hunk 5 above)
Comment (gordonkoehn) [reply src/pyggdrasil/analyze/_utils.py:74]: Oh - yes of course did so.
Isssue at 
- #87
